src/domains/ValuesDomain.js

```diff
@@ -351,15 +355,15 @@ export default class ValuesDomain {
       }
 
       // 4. Return the result of negating oldValue; that is, compute a Number with the same magnitude but opposite sign.
-      return new NumberValue(realm, -oldValue);
+      return value instanceof IntegralValue ? new IntegralValue(realm, -oldValue) : new NumberValue(realm, -oldValue);
```

Comment:
Instead of doing a bunch of tests to determine which constructor to use, why not create a single factory function that simply checks if its number argument is actually an integer?

Reply:
IMO, using value to judge types is not a good choice

Reply:
For concrete values it is a very reasonable thing to do. The main thing is to make the code easier to read and to make it more difficult to forget to maintain the invariant that an integer value is always wrapped in a IntegralValue.